src/geode/mesh/core/regular_grid.cpp

```diff
@@ -0,0 +1,234 @@
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ */
+
+#include <geode/mesh/core/regular_grid.h>
+
+#include <geode/basic/attribute_manager.h>
+#include <geode/basic/bitsery_archive.h>
+#include <geode/basic/pimpl_impl.h>
+
+#include <geode/geometry/point.h>
+namespace geode
+{
+    template < index_t dimension >
+    class RegularGrid< dimension >::Impl
+    {
+    public:
+        Impl() = default;
+
+        Impl( Point< dimension > origin,
+            std::array< index_t, dimension > cells_number,
+            std::array< double, dimension > cells_size )
+            : origin_( std::move( origin ) ),
+              cells_number_( std::move( cells_number ) ),
+              cells_size_( std::move( cells_size ) )
+        {
+            index_t nb_cells{ 1 };
+            for( const auto c : cells_number )
+            {
+                nb_cells *= c;
+            }
+            cell_attribute_manager_.resize( nb_cells );
+        }
+
+        AttributeManager& cell_attribute_manager() const
+        {
+            return cell_attribute_manager_;
+        }
+
+        index_t nb_cells( index_t direction ) const
+        {
+            return cells_number_.at( direction );
+        }
+
+        index_t cell_index( const Index& index ) const
+        {
+            index_t cell_id{ 0 };
+            for( const auto d : Range{ dimension } )
+            {
+                OPENGEODE_ASSERT( index[d] < cells_number_[d],
+                    "[RegularGrid::point] Invalid index" );
+                index_t offset{ 1 };
+                for( const auto d2 : Range{ d } )
+                {
+                    offset *= cells_number_[d2];
+                }
+                cell_id += offset * index[d];
+            }
+            return cell_id;
+        }
+
+        Point< dimension > point( const Index& index ) const
+        {
+            Point< dimension > translation;
+            for( const auto d : Range{ dimension } )
+            {
+                OPENGEODE_ASSERT( index[d] < cells_number_[d],
+                    "[RegularGrid::point] Invalid index" );
+                translation.set_value( d, cells_size_[d] * index[d] );
+            }
+            return origin_ + translation;
+        }
+
+        const Point< dimension >& origin() const
+        {
+            return origin_;
+        }
+
+        absl::optional< Index > cell( const Point< dimension >& query ) const
+        {
+            Index index;
+            for( const auto d : Range{ dimension } )
+            {
+                const auto value =
+                    std::floor( ( query.value( d ) - origin_.value( d ) )
+                                / cells_size_[d] );
+                if( value < 0 || value >= cells_number_[d] )
+                {
+                    return absl::nullopt;
+                }
+                index[d] = static_cast< index_t >( value );

```

Comment:
Should the cast be done after or before comparison with integer values?

Reply:
If it is done before, you have issue because value can be negative

Reply:
True!